vortex/src/main/java/org/apache/iceberg/vortex/Vortex.java

```diff
@@ -0,0 +1,136 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.vortex;
+
+import dev.vortex.api.DType;
+import java.util.Map;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.InternalData;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.data.Record;
+import org.apache.iceberg.data.vortex.GenericVortexReader;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.datafile.DataFileServiceRegistry;
+import org.apache.iceberg.mapping.NameMapping;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+/** Entrypoint to working with Vortex {@link FileFormat formatted} content files. */
+public final class Vortex {
+  private Vortex() {}
+
+  public static void register() {
+    // Register generic Vortex reader.
+    DataFileServiceRegistry.registerReader(
+        FileFormat.VORTEX,
+        Record.class.getName(),
+        inputFile -> read(inputFile).readerFunction(GenericVortexReader::buildReader));
+  }
+
+  public static ReadBuilder read(InputFile inputFile) {
+    return new ReadBuilder(inputFile);
+  }
+
+  public interface ReaderFunction<R> {
+    VortexRowReader<R> read(Schema schema, DType fileSchema, Map<Integer, ?> idToConstant);
+  }
+
+  public static final class ReadBuilder
+      implements InternalData.ReadBuilder,
+          org.apache.iceberg.io.datafile.ReadBuilder<ReadBuilder, Object> {
+    private final InputFile inputFile;
+    private Schema schema;
+    private ReaderFunction<?> readerFunction;
+    private Map<Integer, ?> idToConstant;
+
+    ReadBuilder(InputFile inputFile) {
+      this.inputFile = inputFile;
+    }
+
+    @Override
+    public ReadBuilder project(Schema projectedSchema) {
+      this.schema = projectedSchema;
+      return this;
+    }
+
+    @Override
+    public ReadBuilder set(String key, String value) {
+      // No-op.
+      return this;
+    }
+
+    @Override
+    public ReadBuilder split(long newStart, long newLength) {
+      // TODO(aduffy): support splitting, need to expose split size configuration over the Rust FFI.
+      throw new UnsupportedOperationException("Splitting is not supported yet.");
+    }
+
+    @Override
+    public ReadBuilder reuseContainers() {
+      // No-op.
+      return this;
+    }
+
+    @Override
+    public ReadBuilder reuseContainers(boolean newReuseContainers) {
+      // This is a no-op.
+      return this;
+    }
+
+    @Override
+    public ReadBuilder idToConstant(Map<Integer, ?> newIdConstant) {
+      this.idToConstant = newIdConstant;
+      return this;
+    }
+
+    @Override
+    public ReadBuilder withNameMapping(NameMapping newNameMapping) {
+      // TODO(aduffy): is this for field renames? Figure out how to patch this through.
+      return this;
+    }
```

Comment:
Iceberg uses column id to map the Iceberg table columns to the data file columns.
New data files written by Iceberg contain the Iceberg column Id in the data file metadata.
When files generated by external writers (For example: migration use-cases for old Hive tables) then column name based mapping is used. This mapping is driven by the `NameMapping` provided here.